onnxruntime/core/framework/session_state_initializer.cc

```diff
@@ -225,27 +241,71 @@ common::Status DeserializeTensorProto(const ONNX_NAMESPACE::TensorProto& tensor_
   mlvalue.Init(p_tensor.release(),
                DataTypeImpl::GetType<Tensor>(),
                DataTypeImpl::GetType<Tensor>()->GetDeleteFunc());
-
   return common::Status::OK();
 }
 
-static common::Status PlanTensor(MLValuePatternPlanner& planner, const MLValueNameIdxMap& mlvalue_name_idx_map, const std::string& name, const ONNX_NAMESPACE::TensorProto& tensor_proto) {
-  int mlvalue_index;
-  ORT_RETURN_IF_ERROR(mlvalue_name_idx_map.GetIdx(name, mlvalue_index));
-  size_t len;
-  Status st = utils::GetSizeInBytesFromTensorProto<256>(tensor_proto, &len);
-  if (st.Code() == common::NOT_IMPLEMENTED) return Status::OK();
-  if (!st.IsOK()) return st;
-  return planner.TraceAllocation(mlvalue_index, len);
+static common::Status AllocatePlannedBuffers(const MemoryPatternGroup& mem_patterns,
+                                             const ExecutionProviders& exec_providers,
+                                             std::map<OrtAllocatorInfo, BufferUniquePtr>& weights_buffers) {
+  for (size_t i = 0; i < mem_patterns.locations.size(); i++) {
+    auto& location = mem_patterns.locations[i];
+    ORT_ENFORCE(weights_buffers.find(location) == weights_buffers.end(), "Existing entry in weights buffer for ",
+                location.name);
+
+    auto alloc = utils::GetAllocator(exec_providers, location);
+    if (!alloc)
+      return Status(common::ONNXRUNTIME, common::FAIL, "Failed to get allocator for location: " + location.ToString());
+
+    if (mem_patterns.patterns[i].PeakSize() > 0) {
+      void* buffer = alloc->Alloc(mem_patterns.patterns[i].PeakSize());
+      auto kvp = weights_buffers.insert(std::make_pair(location, BufferUniquePtr(buffer, alloc)));
+      if (!kvp.second) {
+        return Status(common::ONNXRUNTIME, common::FAIL, "duplicated location");
+      }
+    }
+  }
+  return Status::OK();
+}
+
+common::Status GetPreallocateBuffer(const MemoryPatternGroup& mem_patterns, const OrtAllocatorInfo& location,
+                                    int mlvalue_index,
+                                    const std::map<OrtAllocatorInfo, BufferUniquePtr>& weights_buffers,
+                                    const char* name, void** p, size_t* len) {
+  auto pattern = mem_patterns.GetPatterns(location);
+  if (pattern == nullptr) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Mem pattern for initializer ", name, " is not found");
+  }
```

Comment:
Is the failure for 'location' or 'name'?

Reply:
Indeed it's name.

Instead of having a map of map(name->pattern),

we build a  map(location, map(name, pattern)) upon it.
First, we get the location by name, then search the location. So, if the map doesn't have an entry of the location, it means we didn't insert the tensor with the name.




---
In reply to: [261404828](https://github.com/Microsoft/onnxruntime/pull/520#discussion_r261404828) [](ancestors = 261404828)